File: 224.cpp
#include<iostream>
#include<vector>
#include<string>
#include<stack>

using namespace std;

class Solution {
    public:
    int calculate(string s) {
        stack<int> op;
        stack<int> help;
        int number=0;
        int sign=1;
        int res=0;
        for(auto c:s){
            if(isdigit(c)){
                number=number*10+c-'0';
            }
            else{
                res+=number*sign;
                number=0;
                if(c=='+')
                sign=1;
                else if(c=='-')
                sign=-1;
                else if(c=='('){
                    op.push(sign);
                    sign=1;//reset to 1, have a brand new expression.
                    help.push(res);
                    res=0;
                }
                else if(c==')'){
                    res=help.top()+op.top()*res;
                    help.pop();
                    op.pop();
                }
            }
        }
        res+=sign*number;
        return res;
    }
};

/*
 1-(2-(1+2))
 1 2
 - -
 res=-1
 1+-(-1)
 2
 
 */

int main(){

}

File: Python/224.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        signs = []
        nums = []
        res = n = 0
        sign = 1
        for i, x in enumerate(s):
            if x.isdigit():
                n = n*10 + int(x)
                continue
            res = res + sign*n
            n = 0
            
            if x == "+":
                sign = 1
            elif x == "-":
                sign = -1
            elif x == "(":
                nums.append(res)
                signs.append(sign)
                sign = 1
                res = 0
            elif x == ")":
                res = nums[-1] + signs[-1]*res
                nums.pop()
                signs.pop()
                
        return res + n*sign
